// thirdweb/dist/cjs/react/core/hooks/useBridgeQuote.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useBridgeQuote = useBridgeQuote;
const react_query_1 = require("@tanstack/react-query");
const Buy = require("../../../bridge/Buy.js");
const Transfer = require("../../../bridge/Transfer.js");
const address_js_1 = require("../../../utils/address.js");
function useBridgeQuote({ originToken, destinationToken, destinationAmount, client, enabled = true, }) {
    return (0, react_query_1.useQuery)({
        enabled: enabled && !!originToken && !!destinationToken && !!destinationAmount,
        queryFn: async () => {
            // if ssame token and chain, use transfer
            if ((0, address_js_1.checksumAddress)(originToken.address) ===
                (0, address_js_1.checksumAddress)(destinationToken.address) &&
                originToken.chainId === destinationToken.chainId) {
                const transfer = await Transfer.prepare({
                    amount: destinationAmount,
                    chainId: originToken.chainId,
                    client,
                    receiver: destinationToken.address,
                    sender: originToken.address,
                    tokenAddress: originToken.address,
                });
                return transfer;
            }
            const quote = await Buy.quote({
                amount: destinationAmount,
                client,
                destinationChainId: destinationToken.chainId,
                destinationTokenAddress: destinationToken.address,
                originChainId: originToken.chainId,
                originTokenAddress: originToken.address,
            });
            return quote;
        },
        queryKey: [
            "bridge-quote",
            originToken.chainId,
            originToken.address,
            destinationToken.chainId,
            destinationToken.address,
            destinationAmount.toString(),
        ],
        refetchInterval: 60000, // 30 seconds
        retry: 3, // 1 minute
        staleTime: 30000,
    });
}
