// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/pay-transactions/SwapDetailsScreen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwapTxDetailsTable = SwapTxDetailsTable;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_icons_1 = require("@radix-ui/react-icons");
const utils_js_1 = require("../../../../../../../chains/utils.js");
const url_js_1 = require("../../../../../../../utils/url.js");
const index_js_1 = require("../../../../../../core/design-system/index.js");
const useChainQuery_js_1 = require("../../../../../../core/hooks/others/useChainQuery.js");
const basic_js_1 = require("../../../../components/basic.js");
const buttons_js_1 = require("../../../../components/buttons.js");
const Spacer_js_1 = require("../../../../components/Spacer.js");
const text_js_1 = require("../../../../components/text.js");
const WalletRow_js_1 = require("../swap/WalletRow.js");
const statusMeta_js_1 = require("./statusMeta.js");
const TokenInfoRow_js_1 = require("./TokenInfoRow.js");
function SwapTxDetailsTable(props) {
    let uiData;
    let showStatusRow = true;
    if (props.type === "status") {
        const status = props.status;
        if (props.hideStatusRow) {
            showStatusRow = false;
        }
        const isPartialSuccess = status.status === "COMPLETED" && status.subStatus === "PARTIAL_SUCCESS";
        uiData = {
            destinationTxHash: status.destination?.transactionHash,
            estimatedDuration: status.quote.estimated.durationSeconds || 0,
            fromAddress: status.fromAddress,
            fromToken: {
                address: status.quote.fromToken.tokenAddress,
                amount: status.quote.fromAmount,
                chainId: status.quote.fromToken.chainId,
                symbol: status.quote.fromToken.symbol || "",
            },
            gotToken: status.destination
                ? {
                    address: status.destination.token.tokenAddress,
                    amount: status.destination.amount,
                    chainId: status.destination.token.chainId,
                    symbol: status.destination.token.symbol || "",
                }
                : undefined,
            isPartialSuccess,
            quotedToToken: {
                address: status.quote.toToken.tokenAddress,
                amount: status.quote.toAmount,
                chainId: status.quote.toToken.chainId,
                symbol: status.quote.toToken.symbol || "",
            },
            sourceTxHash: status.source?.transactionHash,
            statusMeta: (0, statusMeta_js_1.getBuyWithCryptoStatusMeta)(status),
            toAddress: status.toAddress,
        };
    }
    else {
        const quote = props.quote;
        uiData = {
            estimatedDuration: quote.swapDetails.estimated.durationSeconds || 0,
            fromAddress: quote.swapDetails.fromAddress,
            fromToken: {
                address: quote.swapDetails.fromToken.tokenAddress,
                amount: quote.swapDetails.fromAmount,
                chainId: quote.swapDetails.fromToken.chainId,
                symbol: quote.swapDetails.fromToken.symbol || "",
            },
            isPartialSuccess: false,
            quotedToToken: {
                address: quote.swapDetails.toToken.tokenAddress,
                amount: quote.swapDetails.toAmount,
                chainId: quote.swapDetails.toToken.chainId,
                symbol: quote.swapDetails.toToken.symbol || "",
            },
            toAddress: quote.swapDetails.toAddress,
        };
    }
    const { client } = props;
    const { fromToken, quotedToToken: toToken, statusMeta, sourceTxHash, destinationTxHash, isPartialSuccess, gotToken, } = uiData;
    const fromChainId = fromToken.chainId;
    const toChainId = toToken.chainId;
    const fromChainName = (0, useChainQuery_js_1.useChainName)((0, utils_js_1.getCachedChain)(fromChainId));
    const fromChainExplorers = (0, useChainQuery_js_1.useChainExplorers)((0, utils_js_1.getCachedChain)(fromChainId));
    const toChainName = (0, useChainQuery_js_1.useChainName)((0, utils_js_1.getCachedChain)(toChainId));
    const toChainExplorers = (0, useChainQuery_js_1.useChainExplorers)((0, utils_js_1.getCachedChain)(toChainId));
    const lineSpacer = ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(basic_js_1.Line, {}), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" })] }));
    return ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(TokenInfoRow_js_1.TokenInfoRow, { chainId: fromToken.chainId, client: client, label: "Paid", tokenAddress: fromToken.address, tokenAmount: fromToken.amount, tokenSymbol: fromToken.symbol || "" }), lineSpacer, isPartialSuccess && gotToken ? (
            // Expected + Got
            (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(TokenInfoRow_js_1.TokenInfoRow, { chainId: toToken.chainId, client: client, label: isPartialSuccess ? "Expected" : "Received", tokenAddress: toToken.address, tokenAmount: toToken.amount, tokenSymbol: toToken.symbol || "" }), lineSpacer, (0, jsx_runtime_1.jsx)(TokenInfoRow_js_1.TokenInfoRow, { chainId: gotToken.chainId, client: client, label: "Got", tokenAddress: gotToken.address, tokenAmount: gotToken.amount, tokenSymbol: gotToken.symbol || "" })] })) : (
            // Receive
            (0, jsx_runtime_1.jsx)(TokenInfoRow_js_1.TokenInfoRow, { chainId: toToken.chainId, client: client, label: "Received", tokenAddress: toToken.address, tokenAmount: toToken.amount, tokenSymbol: toToken.symbol || "" })), lineSpacer, (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", style: {
                    justifyContent: "space-between",
                }, children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "sm", children: "Recipient" }), (0, jsx_runtime_1.jsx)(WalletRow_js_1.WalletRow, { address: uiData.toAddress, client: client, iconSize: "sm" })] }), statusMeta && showStatusRow && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [lineSpacer, (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", style: {
                            justifyContent: "space-between",
                        }, children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "sm", children: "Status" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { center: "y", flex: "column", gap: "3xs", style: {
                                    alignItems: "flex-end",
                                }, children: (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: statusMeta.color, size: "sm", children: statusMeta.status }) })] })] })), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), fromChainExplorers.explorers?.[0]?.url && sourceTxHash && ((0, jsx_runtime_1.jsxs)(buttons_js_1.ButtonLink, { fullWidth: true, gap: "xs", href: (0, url_js_1.formatExplorerTxUrl)(fromChainExplorers.explorers[0]?.url, sourceTxHash), style: {
                    fontSize: index_js_1.fontSize.sm,
                    padding: index_js_1.spacing.sm,
                }, target: "_blank", variant: "outline", children: ["View on ", fromChainName.name, " Explorer", (0, jsx_runtime_1.jsx)(react_icons_1.ExternalLinkIcon, { height: index_js_1.iconSize.sm, width: index_js_1.iconSize.sm })] })), destinationTxHash &&
                sourceTxHash !== destinationTxHash &&
                toChainExplorers?.explorers?.[0]?.url && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsxs)(buttons_js_1.ButtonLink, { fullWidth: true, gap: "xs", href: (0, url_js_1.formatExplorerTxUrl)(toChainExplorers.explorers[0]?.url, destinationTxHash), style: {
                            fontSize: index_js_1.fontSize.sm,
                            padding: index_js_1.spacing.sm,
                        }, target: "_blank", variant: "outline", children: ["View on ", toChainName.name, " Explorer", (0, jsx_runtime_1.jsx)(react_icons_1.ExternalLinkIcon, { height: index_js_1.iconSize.sm, width: index_js_1.iconSize.sm })] })] }))] }));
}
