// thirdweb/dist/esm/react/web/ui/Bridge/CheckoutWidget.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { useQuery } from "@tanstack/react-query";
import { NATIVE_TOKEN_ADDRESS } from "../../../../constants/addresses.js";
import { getToken } from "../../../../pay/convert/get-token.js";
import { checksumAddress } from "../../../../utils/address.js";
import { stringify } from "../../../../utils/json.js";
import { CustomThemeProvider } from "../../../core/design-system/CustomThemeProvider.js";
import { useConnectLocale } from "../ConnectWallet/locale/getConnectLocale.js";
import { EmbedContainer } from "../ConnectWallet/Modal/ConnectEmbed.js";
import { DynamicHeight } from "../components/DynamicHeight.js";
import { Spinner } from "../components/Spinner.js";
import { BridgeOrchestrator } from "./BridgeOrchestrator.js";
import { UnsupportedTokenScreen } from "./UnsupportedTokenScreen.js";
/**
 * Widget a prebuilt UI for purchasing a specific token.
 *
 * @param props - Props of type [`CheckoutWidgetProps`](https://portal.thirdweb.com/references/typescript/v5/CheckoutWidgetProps) to configure the CheckoutWidget component.
 *
 * @example
 * ### Default configuration
 *
 * By default, the `CheckoutWidget` component will allows users to fund their wallets with crypto or fiat on any of the supported chains..
 *
 * ```tsx
 * <CheckoutWidget
 *   client={client}
 *  />
 * ```
 *
 * ### Customize the UI
 *
 * You can customize the UI of the `CheckoutWidget` component by passing a custom theme object to the `theme` prop.
 *
 * ```tsx
 * <CheckoutWidget
 *   client={client}
 *   theme={darkTheme({
 *     colors: {
 *       modalBg: "red",
 *     },
 *   })}
 * />
 * ```
 *
 * Refer to the [`Theme`](https://portal.thirdweb.com/references/typescript/v5/Theme) type for more details.
 *
 * ### Update the Title
 *
 * You can update the title of the widget by passing a `title` prop to the `CheckoutWidget` component.
 *
 * ```tsx
 * <CheckoutWidget
 *   client={client}
 *   title="Checkout ETH"
 * />
 * ```
 *
 * ### Configure the wallet connection
 *
 * You can customize the wallet connection flow by passing a `connectOptions` object to the `CheckoutWidget` component.
 *
 * ```tsx
 * <CheckoutWidget
 *   client={client}
 *   connectOptions={{
 *     connectModal: {
 *       size: 'compact',
 *       title: "Sign in",
 *     }
 *   }}
 * />
 * ```
 *
 * Refer to the [`CheckoutWidgetConnectOptions`](https://portal.thirdweb.com/references/typescript/v5/CheckoutWidgetConnectOptions) type for more details.
 *
 * @bridge
 * @beta
 * @react
 */
export function CheckoutWidget(props) {
    const localeQuery = useConnectLocale(props.locale || "en_US");
    const theme = props.theme || "dark";
    const bridgeDataQuery = useQuery({
        queryFn: async () => {
            const token = await getToken(props.client, checksumAddress(props.tokenAddress || NATIVE_TOKEN_ADDRESS), props.chain.id).catch((err) => err.message.includes("not supported") ? undefined : Promise.reject(err));
            if (!token) {
                return {
                    chain: props.chain,
                    tokenAddress: checksumAddress(props.tokenAddress || NATIVE_TOKEN_ADDRESS),
                    type: "unsupported_token",
                };
            }
            return {
                data: {
                    metadata: {
                        description: props.description,
                        image: props.image,
                        title: props.name,
                    },
                    mode: "direct_payment",
                    paymentInfo: {
                        amount: props.amount,
                        feePayer: props.feePayer === "seller" ? "receiver" : "sender",
                        sellerAddress: props.seller,
                        token, // User is sender, seller is receiver
                    },
                },
                type: "success",
            };
        },
        queryKey: ["bridgeData", stringify(props)],
    });
    let content = null;
    if (!localeQuery.data || bridgeDataQuery.isLoading) {
        content = (_jsx("div", { style: {
                alignItems: "center",
                display: "flex",
                justifyContent: "center",
                minHeight: "350px",
            }, children: _jsx(Spinner, { color: "secondaryText", size: "xl" }) }));
    }
    else if (bridgeDataQuery.data?.type === "unsupported_token") {
        // Show unsupported token screen
        content = _jsx(UnsupportedTokenScreen, { chain: bridgeDataQuery.data.chain });
    }
    else if (bridgeDataQuery.data?.type === "success") {
        // Show normal bridge orchestrator
        content = (_jsx(BridgeOrchestrator, { client: props.client, connectLocale: localeQuery.data, connectOptions: props.connectOptions, onCancel: () => {
                props.onCancel?.();
            }, onComplete: () => {
                props.onSuccess?.();
            }, onError: (err) => {
                props.onError?.(err);
            }, paymentLinkId: props.paymentLinkId, presetOptions: props.presetOptions, purchaseData: props.purchaseData, receiverAddress: props.seller, uiOptions: bridgeDataQuery.data.data }));
    }
    return (_jsx(CustomThemeProvider, { theme: theme, children: _jsx(EmbedContainer, { className: props.className, modalSize: "compact", style: props.style, children: _jsx(DynamicHeight, { children: content }) }) }));
}
