// thirdweb/dist/esm/react/web/ui/Bridge/TransactionPayment.js
"use client";
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { getAddress, shortenAddress, } from "../../../../utils/address.js";
import { useCustomTheme } from "../../../core/design-system/CustomThemeProvider.js";
import { fontSize, spacing, } from "../../../core/design-system/index.js";
import { useChainMetadata } from "../../../core/hooks/others/useChainQuery.js";
import { useTransactionDetails } from "../../../core/hooks/useTransactionDetails.js";
import { useActiveAccount } from "../../../core/hooks/wallets/useActiveAccount.js";
import { ConnectButton } from "../ConnectWallet/ConnectButton.js";
import { PoweredByThirdweb } from "../ConnectWallet/PoweredByTW.js";
import { Container, Line } from "../components/basic.js";
import { Button } from "../components/buttons.js";
import { ChainName } from "../components/ChainName.js";
import { Spacer } from "../components/Spacer.js";
import { Text } from "../components/text.js";
import { ChainIcon } from "./common/TokenAndChain.js";
import { WithHeader } from "./common/WithHeader.js";
export function TransactionPayment({ uiOptions, client, onContinue, connectOptions, }) {
    const theme = useCustomTheme();
    const activeAccount = useActiveAccount();
    // Get chain metadata for native currency symbol
    const chainMetadata = useChainMetadata(uiOptions.transaction.chain);
    // Use the extracted hook for transaction details
    const transactionDataQuery = useTransactionDetails({
        client,
        transaction: uiOptions.transaction,
    });
    const contractName = transactionDataQuery.data?.contractMetadata?.name || "Unknown Contract";
    const functionName = transactionDataQuery.data?.functionInfo?.functionName || "Contract Call";
    const isLoading = transactionDataQuery.isLoading || chainMetadata.isLoading;
    const buttonLabel = `Execute ${functionName}`;
    if (isLoading) {
        return (_jsxs(WithHeader, { client: client, defaultTitle: "Transaction", uiOptions: uiOptions, children: [_jsx(SkeletonHeader, { theme: theme }), _jsx(Spacer, { y: "md" }), _jsx(Line, {}), _jsx(Spacer, { y: "md" }), _jsx(SkeletonRow, { theme: theme, width: "60%" }), _jsx(Spacer, { y: "xs" }), _jsx(SkeletonRow, { theme: theme, width: "40%" }), _jsx(Spacer, { y: "xs" }), _jsx(SkeletonRow, { theme: theme, width: "50%" }), _jsx(Spacer, { y: "xs" }), _jsx(SkeletonRow, { theme: theme, width: "45%" }), _jsx(Spacer, { y: "xs" }), _jsx(SkeletonRow, { theme: theme, width: "55%" }), _jsx(Spacer, { y: "md" }), _jsx(Line, {}), _jsx(Spacer, { y: "lg" }), _jsx("div", { style: {
                        backgroundColor: theme.colors.skeletonBg,
                        borderRadius: spacing.md,
                        height: "48px",
                        width: "100%",
                    } }), _jsx(Spacer, { y: "md" }), _jsx(PoweredByThirdweb, {}), _jsx(Spacer, { y: "md" })] }));
    }
    return (_jsxs(WithHeader, { client: client, defaultTitle: "Transaction", uiOptions: uiOptions, children: [_jsxs(Container, { center: "y", flex: "row", gap: "3xs", style: {
                    justifyContent: "space-between",
                }, children: [_jsx(Text, { color: "primaryText", size: "xl", weight: 700, children: transactionDataQuery.data?.usdValueDisplay ||
                            transactionDataQuery.data?.txCostDisplay }), _jsx(Text, { color: "secondaryText", size: "md", style: {
                            backgroundColor: theme.colors.tertiaryBg,
                            borderRadius: spacing.sm,
                            fontFamily: "monospace",
                            padding: `${spacing.xs} ${spacing.sm}`,
                            textAlign: "right",
                        }, children: functionName })] }), _jsx(Spacer, { y: "md" }), _jsx(Line, {}), _jsx(Spacer, { y: "md" }), _jsxs(Container, { flex: "row", style: {
                    alignItems: "center",
                    justifyContent: "space-between",
                }, children: [_jsx(Text, { color: "secondaryText", size: "sm", children: "Contract" }), _jsx(Text, { color: "primaryText", size: "sm", children: contractName })] }), _jsx(Spacer, { y: "xs" }), _jsxs(Container, { flex: "row", style: {
                    alignItems: "center",
                    justifyContent: "space-between",
                }, children: [_jsx(Text, { color: "secondaryText", size: "sm", children: "Address" }), _jsx("a", { href: `https://thirdweb.com/${uiOptions.transaction.chain.id}/${uiOptions.transaction.to}`, rel: "noopener noreferrer", style: {
                            color: theme.colors.accentText,
                            fontFamily: "monospace",
                            fontSize: fontSize.sm,
                            textDecoration: "none",
                        }, target: "_blank", children: shortenAddress(uiOptions.transaction.to) })] }), _jsx(Spacer, { y: "xs" }), _jsxs(Container, { flex: "row", style: {
                    alignItems: "center",
                    justifyContent: "space-between",
                }, children: [_jsx(Text, { color: "secondaryText", size: "sm", children: "Network" }), _jsxs(Container, { center: "y", flex: "row", gap: "3xs", children: [_jsx(ChainIcon, { chain: uiOptions.transaction.chain, client: client, size: "xs" }), _jsx(ChainName, { chain: uiOptions.transaction.chain, client: client, color: "primaryText", short: true, size: "sm", style: {
                                    fontFamily: "monospace",
                                } })] })] }), _jsx(Spacer, { y: "xs" }), transactionDataQuery.data?.txCostDisplay && (_jsxs(_Fragment, { children: [_jsxs(Container, { flex: "row", style: {
                            alignItems: "center",
                            justifyContent: "space-between",
                        }, children: [_jsx(Text, { color: "secondaryText", size: "sm", children: "Cost" }), _jsx(Text, { color: "primaryText", size: "sm", style: {
                                    fontFamily: "monospace",
                                }, children: transactionDataQuery.data?.txCostDisplay })] }), _jsx(Spacer, { y: "xs" })] })), transactionDataQuery.data?.gasCostDisplay && (_jsxs(_Fragment, { children: [_jsxs(Container, { flex: "row", style: {
                            alignItems: "center",
                            justifyContent: "space-between",
                        }, children: [_jsx(Text, { color: "secondaryText", size: "sm", children: "Network fees" }), _jsx(Text, { color: "primaryText", size: "sm", style: {
                                    fontFamily: "monospace",
                                }, children: transactionDataQuery.data?.gasCostDisplay })] }), _jsx(Spacer, { y: "md" })] })), _jsx(Line, {}), _jsx(Spacer, { y: "lg" }), activeAccount ? (_jsx(Button, { fullWidth: true, onClick: () => {
                    if (transactionDataQuery.data?.tokenInfo) {
                        onContinue(transactionDataQuery.data.totalCost, transactionDataQuery.data.tokenInfo, getAddress(activeAccount.address));
                    }
                }, style: {
                    fontSize: fontSize.md,
                    padding: `${spacing.sm} ${spacing.md}`,
                }, variant: "primary", children: buttonLabel })) : (_jsx(ConnectButton, { client: client, connectButton: {
                    label: buttonLabel,
                }, theme: theme, ...connectOptions })), _jsx(Spacer, { y: "md" }), _jsx(PoweredByThirdweb, {}), _jsx(Spacer, { y: "lg" })] }));
}
const SkeletonHeader = (props) => (_jsxs(Container, { center: "y", flex: "row", gap: "3xs", style: {
        justifyContent: "space-between",
    }, children: [_jsx("div", { style: {
                backgroundColor: props.theme.colors.skeletonBg,
                borderRadius: spacing.xs,
                height: "32px",
                width: "80px",
            } }), _jsx("div", { style: {
                backgroundColor: props.theme.colors.skeletonBg,
                borderRadius: spacing.sm,
                height: "24px",
                width: "120px",
            } })] }));
// Skeleton component for loading state
const SkeletonRow = ({ width = "100%", theme, }) => (_jsxs(Container, { flex: "row", style: {
        alignItems: "center",
        justifyContent: "space-between",
    }, children: [_jsx("div", { style: {
                backgroundColor: theme.colors.skeletonBg,
                borderRadius: spacing.xs,
                height: "16px",
                width: "30%",
            } }), _jsx("div", { style: {
                backgroundColor: theme.colors.skeletonBg,
                borderRadius: spacing.xs,
                height: "16px",
                width,
            } })] }));
