// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/PayWIthCreditCard.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import styled from "@emotion/styled";
import { ChevronDownIcon } from "@radix-ui/react-icons";
import { formatNumber } from "../../../../../../utils/formatNumber.js";
import { fontSize, iconSize, radius, spacing, } from "../../../../../core/design-system/index.js";
import { Container } from "../../../components/basic.js";
import { Button } from "../../../components/buttons.js";
import { Skeleton } from "../../../components/Skeleton.js";
import { Text } from "../../../components/text.js";
import { getFiatIcon } from "./fiat/currencies.js";
/**
 * Shows an amount "value" and renders the selected token and chain
 * It also renders the buttons to select the token and chain
 * It also renders the balance of active wallet for the selected token in selected chain
 * @internal
 */
export function PayWithCreditCard(props) {
    return (_jsxs(Container, { bg: "tertiaryBg", borderColor: "borderColor", flex: "row", style: {
            alignItems: "center",
            borderBottom: "none",
            borderBottomLeftRadius: 0,
            borderBottomRightRadius: 0,
            borderRadius: radius.md,
            borderStyle: "solid",
            borderWidth: "1px",
            flexWrap: "nowrap",
            justifyContent: "space-between",
        }, children: [_jsxs(CurrencyButton, { gap: "sm", onClick: props.onSelectCurrency, style: {
                    justifyContent: "flex-start",
                    minHeight: "64px",
                    minWidth: "50%",
                }, variant: "ghost", children: [getFiatIcon(props.currency, "md"), _jsxs(Container, { center: "y", color: "secondaryText", flex: "row", gap: "xxs", children: [_jsx(Text, { color: "primaryText", children: props.currency.shorthand }), _jsx(ChevronDownIcon, { height: iconSize.sm, width: iconSize.sm })] })] }), _jsx("div", { style: {
                    alignItems: "flex-end",
                    display: "flex",
                    flexDirection: "column",
                    flexGrow: 1,
                    flexShrink: 1,
                    gap: spacing.xxs,
                    justifyContent: "center",
                    overflow: "hidden",
                    paddingRight: spacing.sm,
                    textOverflow: "ellipsis",
                    whiteSpace: "nowrap",
                }, children: props.isLoading ? (_jsx(Skeleton, { height: fontSize.lg, width: "100px" })) : (_jsx(Text, { color: props.value ? "primaryText" : "secondaryText", size: "lg", children: props.value
                        ? `${props.currency.symbol}${formatNumber(Number(props.value), 2)}`
                        : "--" })) })] }));
}
const CurrencyButton = /* @__PURE__ */ styled(Button)(() => {
    return {
        "&[disabled]:hover": {
            borderColor: "transparent",
        },
    };
});
