// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/TransactionModeScreen.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { trackPayEvent } from "../../../../../../analytics/track/pay.js";
import { NATIVE_TOKEN_ADDRESS } from "../../../../../../constants/addresses.js";
import { formatNumber } from "../../../../../../utils/formatNumber.js";
import { toTokens } from "../../../../../../utils/units.js";
import { hasSponsoredTransactionsEnabled } from "../../../../../../wallets/smart/is-smart-wallet.js";
import { useCustomTheme } from "../../../../../core/design-system/CustomThemeProvider.js";
import { fontSize, iconSize, spacing, } from "../../../../../core/design-system/index.js";
import { useChainMetadata } from "../../../../../core/hooks/others/useChainQuery.js";
import { useWalletBalance } from "../../../../../core/hooks/others/useWalletBalance.js";
import { useActiveAccount } from "../../../../../core/hooks/wallets/useActiveAccount.js";
import { useActiveWallet } from "../../../../../core/hooks/wallets/useActiveWallet.js";
import { ErrorState } from "../../../../wallets/shared/ErrorState.js";
import { LoadingScreen } from "../../../../wallets/shared/LoadingScreen.js";
import { Container, Line, ModalHeader } from "../../../components/basic.js";
import { Button } from "../../../components/buttons.js";
import { ChainIcon } from "../../../components/ChainIcon.js";
import { Img } from "../../../components/Img.js";
import { Skeleton } from "../../../components/Skeleton.js";
import { Spacer } from "../../../components/Spacer.js";
import { TokenIcon } from "../../../components/TokenIcon.js";
import { Text } from "../../../components/text.js";
import { TokenSymbol } from "../../../components/token/TokenSymbol.js";
import { ConnectButton } from "../../ConnectButton.js";
import { OutlineWalletIcon } from "../../icons/OutlineWalletIcon.js";
import { PoweredByThirdweb } from "../../PoweredByTW.js";
import { formatTokenBalance } from "../formatTokenBalance.js";
import { isNativeToken, NATIVE_TOKEN, } from "../nativeToken.js";
import { useTransactionCostAndData } from "./main/useBuyTxStates.js";
import { WalletRow } from "./swap/WalletRow.js";
export function TransactionModeScreen(props) {
    const { payUiOptions, client, payerAccount, supportedDestinations, onContinue, } = props;
    const { data: chainData, error: chainDataError, isLoading: chainDataLoading, refetch: chainDataRefetch, } = useChainMetadata(payUiOptions.transaction.chain);
    const metadata = payUiOptions.metadata;
    const { data: transactionCostAndData, error: transactionCostAndDataError, isLoading: transactionCostAndDataLoading, refetch: transactionCostAndDataRefetch, } = useTransactionCostAndData({
        account: payerAccount,
        supportedDestinations,
        transaction: payUiOptions.transaction,
    });
    const theme = useCustomTheme();
    const activeWallet = useActiveWallet();
    const activeAccount = useActiveAccount();
    const sponsoredTransactionsEnabled = hasSponsoredTransactionsEnabled(activeWallet);
    const balanceQuery = useWalletBalance({
        address: activeAccount?.address,
        chain: payUiOptions.transaction.chain,
        client: props.client,
        tokenAddress: isNativeToken(transactionCostAndData?.token || NATIVE_TOKEN)
            ? undefined
            : transactionCostAndData?.token.address,
    }, {
        enabled: !!transactionCostAndData,
    });
    if (transactionCostAndDataLoading || chainDataLoading) {
        return _jsx(LoadingScreen, {});
    }
    if (!activeAccount) {
        return (_jsx(Container, { center: "both", flex: "row", fullHeight: true, style: {
                minHeight: "350px",
            }, children: _jsxs(Container, { animate: "fadein", children: [_jsx(Spacer, { y: "xxl" }), _jsx(Container, { center: "x", flex: "row", children: _jsx(OutlineWalletIcon, { size: iconSize["3xl"] }) }), _jsx(Spacer, { y: "lg" }), _jsx(Text, { center: true, color: "primaryText", size: "md", children: "Please connect a wallet to continue" }), _jsx(Spacer, { y: "xl" }), _jsx(Container, { center: "x", flex: "row", style: { width: "100%" }, children: _jsx(ConnectButton, { client: client, theme: theme, ...props.connectOptions }) })] }) }));
    }
    if (transactionCostAndDataError || chainDataError) {
        return (_jsx(Container, { center: "both", flex: "row", fullHeight: true, style: {
                minHeight: "350px",
            }, children: _jsx(ErrorState, { onTryAgain: transactionCostAndDataError
                    ? transactionCostAndDataRefetch
                    : chainDataRefetch, title: transactionCostAndDataError?.message ||
                    chainDataError?.message ||
                    "Something went wrong" }) }));
    }
    if (!transactionCostAndData || !chainData) {
        return _jsx(LoadingScreen, {});
    }
    const insufficientFunds = balanceQuery.data &&
        balanceQuery.data.value < transactionCostAndData.transactionValueWei;
    return (_jsxs(Container, { px: "lg", children: [_jsx(Spacer, { y: "lg" }), _jsx(ModalHeader, { title: metadata?.name || "Transaction" }), _jsx(Spacer, { y: "lg" }), _jsxs(Container, { children: [metadata?.image ? (_jsx(Img, { client: client, src: metadata?.image, style: {
                            backgroundColor: theme.colors.tertiaryBg,
                            border: `1px solid ${theme.colors.borderColor}`,
                            borderRadius: spacing.md,
                            width: "100%",
                        } })) : activeAccount ? (_jsxs(Container, { flex: "column", gap: "sm", children: [insufficientFunds && (_jsxs("div", { children: [_jsx(Text, { center: true, color: "danger", multiline: true, size: "xs", children: "Insufficient Funds" }), _jsx(Text, { center: true, multiline: true, size: "xs", children: "Select another token or pay with card." })] })), _jsxs(Container, { flex: "row", style: {
                                    backgroundColor: theme.colors.tertiaryBg,
                                    border: `1px solid ${theme.colors.borderColor}`,
                                    borderRadius: spacing.md,
                                    justifyContent: "space-between",
                                    marginBottom: spacing.sm,
                                    padding: spacing.sm,
                                }, children: [_jsx(WalletRow, { address: activeAccount?.address, client: client, iconSize: "md" }), balanceQuery.data ? (_jsxs(Container, { center: "y", flex: "row", gap: "3xs", children: [_jsx(Text, { color: "secondaryText", size: "xs", weight: 500, children: formatTokenBalance(balanceQuery.data, false) }), _jsx(TokenSymbol, { chain: payUiOptions.transaction.chain, color: "secondaryText", size: "xs", token: transactionCostAndData.token })] })) : (_jsx(Skeleton, { height: fontSize.xs, width: "70px" }))] })] })) : null, _jsx(Spacer, { y: "md" }), _jsxs(Container, { flex: "row", children: [_jsx(Container, { expand: true, flex: "column", children: _jsx(Text, { color: "primaryText", size: "md", weight: 700, children: "Price" }) }), _jsx(Container, { expand: true, children: _jsxs(Container, { center: "y", flex: "row", gap: "xs", style: { justifyContent: "right" }, children: [_jsx(TokenIcon, { chain: payUiOptions.transaction.chain, client: props.client, size: "sm", token: transactionCostAndData.token }), _jsxs(Text, { color: "primaryText", size: "md", weight: 700, children: [String(formatNumber(Number(toTokens(transactionCostAndData.transactionValueWei, transactionCostAndData.decimals)), 6)), " ", transactionCostAndData.token.symbol] })] }) })] }), _jsx(Spacer, { y: "md" }), _jsx(Line, {}), _jsx(Spacer, { y: "md" }), _jsxs(Container, { flex: "row", children: [_jsx(Container, { expand: true, flex: "column", children: _jsx(Text, { color: "secondaryText", size: "xs", children: "Gas Fees" }) }), _jsx(Container, { expand: true, children: _jsx(Container, { center: "y", flex: "row", gap: "xs", style: { justifyContent: "right" }, children: _jsx(Text, { color: sponsoredTransactionsEnabled ? "success" : "primaryText", size: "xs", children: sponsoredTransactionsEnabled
                                            ? "Sponsored"
                                            : `${String(formatNumber(Number(toTokens(transactionCostAndData.gasCostWei, chainData.nativeCurrency.decimals)), 6))} ${chainData.nativeCurrency.symbol}` }) }) })] }), _jsx(Spacer, { y: "sm" }), _jsxs(Container, { flex: "row", children: [_jsx(Container, { expand: true, flex: "column", children: _jsx(Text, { color: "secondaryText", size: "xs", children: "Network" }) }), _jsx(Container, { expand: true, children: _jsxs(Container, { center: "y", flex: "row", gap: "xs", style: { justifyContent: "right" }, children: [_jsx(ChainIcon, { chainIconUrl: chainData.icon?.url, client: props.client, size: "xs" }), _jsx(Text, { color: "secondaryText", size: "xs", style: { textAlign: "right" }, children: chainData.name })] }) })] })] }), _jsx(Spacer, { y: "xl" }), payerAccount ? (_jsx(Button, { fullWidth: true, onClick: () => {
                    let totalCostWei = insufficientFunds
                        ? transactionCostAndData.transactionValueWei -
                            (balanceQuery.data?.value || 0n)
                        : transactionCostAndData.transactionValueWei;
                    if (transactionCostAndData.token.address === NATIVE_TOKEN_ADDRESS &&
                        !sponsoredTransactionsEnabled) {
                        totalCostWei += transactionCostAndData.gasCostWei;
                    }
                    trackPayEvent({
                        amountWei: totalCostWei.toString(),
                        client,
                        event: "choose_payment_method_transaction_mode",
                        toChainId: payUiOptions.transaction.chain.id,
                        toToken: transactionCostAndData.token.address,
                        walletAddress: payerAccount.address,
                        walletType: activeWallet?.id,
                    });
                    onContinue(toTokens(totalCostWei, transactionCostAndData.decimals), payUiOptions.transaction.chain, transactionCostAndData.token);
                }, variant: "accent", children: "Choose Payment Method" })) : (_jsx("div", { children: _jsx(ConnectButton, { ...props.connectOptions, client: client, connectButton: {
                        style: {
                            width: "100%",
                        },
                    }, theme: theme }) })), _jsx(Spacer, { y: "lg" }), payUiOptions.showThirdwebBranding !== false && (_jsxs(_Fragment, { children: [_jsx(PoweredByThirdweb, { link: "https://playground.thirdweb.com/connect/pay?utm_source=ub_text" }), _jsx(Spacer, { y: "sm" })] }))] }));
}
