// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/swap/SwapStatusScreen.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { CheckCircledIcon } from "@radix-ui/react-icons";
import { useQueryClient } from "@tanstack/react-query";
import { useEffect, useRef, useState } from "react";
import { iconSize } from "../../../../../../core/design-system/index.js";
import { useBuyWithCryptoStatus } from "../../../../../../core/hooks/pay/useBuyWithCryptoStatus.js";
import { invalidateWalletBalance } from "../../../../../../core/providers/invalidateWalletBalance.js";
import { Container, ModalHeader } from "../../../../components/basic.js";
import { Button } from "../../../../components/buttons.js";
import { Spacer } from "../../../../components/Spacer.js";
import { Spinner } from "../../../../components/Spinner.js";
import { Text } from "../../../../components/text.js";
import { AccentFailIcon } from "../../../icons/AccentFailIcon.js";
import { SwapTxDetailsTable } from "../pay-transactions/SwapDetailsScreen.js";
export function SwapStatusScreen(props) {
    const { onSuccess } = props;
    const [showDetails, setShowDetails] = useState(false);
    const swapStatus = useBuyWithCryptoStatus({
        chainId: props.fromChain.id,
        client: props.client,
        transactionHash: props.swapTxHash,
    });
    let uiStatus = "pending";
    if (swapStatus.data?.status === "COMPLETED") {
        uiStatus = "success";
    }
    else if (swapStatus.data?.status === "FAILED") {
        uiStatus = "failed";
    }
    if (swapStatus.data?.status === "COMPLETED" &&
        swapStatus.data?.subStatus === "PARTIAL_SUCCESS") {
        uiStatus = "partialSuccess";
    }
    const purchaseCbCalled = useRef(false);
    useEffect(() => {
        if (purchaseCbCalled.current || !onSuccess) {
            return;
        }
        if (swapStatus.data?.status === "COMPLETED") {
            purchaseCbCalled.current = true;
            onSuccess(swapStatus.data);
        }
    }, [onSuccess, swapStatus]);
    const queryClient = useQueryClient();
    const balanceInvalidated = useRef(false);
    useEffect(() => {
        if ((uiStatus === "success" || uiStatus === "partialSuccess") &&
            !balanceInvalidated.current) {
            balanceInvalidated.current = true;
            invalidateWalletBalance(queryClient);
        }
    }, [queryClient, uiStatus]);
    const swapDetails = swapStatus.data && swapStatus.data.status !== "NOT_FOUND" ? (_jsx(SwapTxDetailsTable, { client: props.client, status: swapStatus.data, type: "status" })) : props.quote ? (_jsx(SwapTxDetailsTable, { client: props.client, quote: props.quote, type: "quote" })) : null;
    if (showDetails) {
        return (_jsx(Container, { animate: "fadein", children: _jsxs(Container, { p: "lg", children: [_jsx(ModalHeader, { onBack: () => setShowDetails(false), title: "Transaction Details" }), _jsx(Spacer, { y: "xl" }), swapDetails] }) }));
    }
    return (_jsx(Container, { animate: "fadein", children: _jsxs(Container, { p: "lg", children: [_jsx(ModalHeader, { onBack: props.onBack, title: props.title }), _jsx(Spacer, { y: "sm" }), uiStatus === "success" && (_jsxs(_Fragment, { children: [_jsx(Spacer, { y: "md" }), _jsxs(Container, { center: "x", color: "success", flex: "column", children: [_jsx(CheckCircledIcon, { height: iconSize["3xl"], width: iconSize["3xl"] }), _jsx(Spacer, { y: "sm" }), _jsx(Text, { color: "primaryText", size: "lg", children: "Buy Complete" })] }), _jsx(Spacer, { y: "xl" }), _jsx(Button, { fullWidth: true, onClick: () => setShowDetails(true), variant: "outline", children: "View transaction details" }), _jsx(Spacer, { y: "sm" }), _jsx(Button, { fullWidth: true, onClick: props.onDone, variant: "accent", children: props.transactionMode ? "Continue Transaction" : "Done" })] })), uiStatus === "partialSuccess" &&
                    swapStatus.data?.status !== "NOT_FOUND" &&
                    swapStatus.data?.destination && (_jsxs(_Fragment, { children: [_jsx(Spacer, { y: "lg" }), _jsxs(Container, { center: "x", color: "success", flex: "column", children: [_jsx(AccentFailIcon, { size: iconSize["3xl"] }), _jsx(Spacer, { y: "xl" }), _jsx(Text, { color: "primaryText", size: "lg", children: "Incomplete" }), _jsx(Spacer, { y: "sm" }), _jsxs(Text, { color: "danger", size: "sm", children: ["Expected ", swapStatus.data.quote.toToken.symbol, ", Got", " ", swapStatus.data.destination.token.symbol, " instead"] })] }), _jsx(Spacer, { y: "xl" })] })), uiStatus === "failed" && (_jsxs(_Fragment, { children: [_jsx(Spacer, { y: "xxl" }), _jsxs(Container, { flex: "column", children: [_jsxs(Container, { center: "both", flex: "column", children: [_jsx(AccentFailIcon, { size: iconSize["3xl"] }), _jsx(Spacer, { y: "xl" }), _jsx(Text, { color: "primaryText", size: "lg", children: "Transaction Failed" }), _jsx(Spacer, { y: "sm" }), _jsxs(Text, { size: "sm", children: ["Your transaction ", `couldn't`, " be processed"] })] }), _jsx(Spacer, { y: "xl" }), _jsx(Button, { fullWidth: true, onClick: () => setShowDetails(true), variant: "outline", children: "View transaction details" }), _jsx(Spacer, { y: "sm" }), _jsx(Button, { fullWidth: true, onClick: props.onTryAgain, variant: "accent", children: "Try Again" })] })] })), uiStatus === "pending" && (_jsxs(_Fragment, { children: [_jsx(Spacer, { y: "xl" }), _jsxs(Container, { animate: "fadein", center: "both", flex: "column", children: [_jsx("div", { style: {
                                        alignItems: "center",
                                        display: "flex",
                                        justifyContent: "center",
                                        position: "relative",
                                    }, children: _jsx(Spinner, { color: "accentText", size: "3xl" }) }), _jsx(Spacer, { y: "lg" }), _jsx(Text, { color: "primaryText", size: "lg", children: "Buy Pending" }), _jsx(Spacer, { y: "sm" }), _jsx(Text, { color: "secondaryText", size: "sm", children: "This may take a minute to complete" })] }), _jsx(Spacer, { y: "xl" })] }))] }) }));
}
