// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/swap/TokenSelectorScreen.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import styled from "@emotion/styled";
import { CardStackIcon, ChevronRightIcon, Cross2Icon, } from "@radix-ui/react-icons";
import { trackPayEvent } from "../../../../../../../analytics/track/pay.js";
import { useCustomTheme } from "../../../../../../core/design-system/CustomThemeProvider.js";
import { iconSize, radius, spacing, } from "../../../../../../core/design-system/index.js";
import { useChainMetadata, useChainName, } from "../../../../../../core/hooks/others/useChainQuery.js";
import { useActiveAccount } from "../../../../../../core/hooks/wallets/useActiveAccount.js";
import { useActiveWallet } from "../../../../../../core/hooks/wallets/useActiveWallet.js";
import { useConnectedWallets } from "../../../../../../core/hooks/wallets/useConnectedWallets.js";
import { useDisconnect } from "../../../../../../core/hooks/wallets/useDisconnect.js";
import { LoadingScreen } from "../../../../../wallets/shared/LoadingScreen.js";
import { Container } from "../../../../components/basic.js";
import { Button } from "../../../../components/buttons.js";
import { Spacer } from "../../../../components/Spacer.js";
import { TextDivider } from "../../../../components/TextDivider.js";
import { TokenIcon } from "../../../../components/TokenIcon.js";
import { Text } from "../../../../components/text.js";
import { OutlineWalletIcon } from "../../../icons/OutlineWalletIcon.js";
import { formatTokenBalance } from "../../formatTokenBalance.js";
import { isNativeToken } from "../../nativeToken.js";
import { FiatValue } from "./FiatValue.js";
import { useWalletsAndBalances, } from "./fetchBalancesForWallet.js";
import { WalletRow } from "./WalletRow.js";
export function TokenSelectorScreen(props) {
    const connectedWallets = useConnectedWallets();
    const activeAccount = useActiveAccount();
    const activeWallet = useActiveWallet();
    const chainInfo = useChainMetadata(props.toChain);
    const theme = useCustomTheme();
    const walletsAndBalances = useWalletsAndBalances({
        client: props.client,
        mode: props.mode,
        sourceSupportedTokens: props.sourceSupportedTokens || [],
        toChain: props.toChain,
        toToken: props.toToken,
    });
    if (walletsAndBalances.isLoading ||
        chainInfo.isLoading ||
        !chainInfo.data ||
        !props.sourceSupportedTokens) {
        return _jsx(LoadingScreen, {});
    }
    const filteredWallets = Array.from(walletsAndBalances.data?.entries() || [])
        .filter(([w]) => !props.hiddenWallets?.includes(w.id))
        .filter(([, balances]) => {
        const hasEnoughBalance = balances.some((b) => b.balance.value > 0);
        return hasEnoughBalance;
    });
    return (_jsxs(Container, { animate: "fadein", style: {
            minHeight: "200px",
        }, children: [filteredWallets.length === 0 ? (_jsx(Container, { flex: "column", gap: "xs", py: "lg", children: _jsxs(Text, { center: true, color: "secondaryText", size: "xs", children: ["No suitable payment token found", _jsx("br", {}), "in connected wallets"] }) })) : (_jsxs(Container, { flex: "column", gap: "xs", children: [_jsx(Text, { size: "sm", children: "Select payment token" }), _jsx(Spacer, { y: "xs" })] })), _jsx(Container, { scrollY: true, style: {
                    maxHeight: "350px",
                }, children: _jsxs(Container, { flex: "column", gap: "sm", children: [filteredWallets.map(([w, balances]) => {
                            const address = w.address;
                            const wallet = connectedWallets.find((w) => w.getAccount()?.address === address);
                            if (!wallet)
                                return null;
                            return (_jsx(WalletRowWithBalances, { address: address, balances: balances, client: props.client, onClick: (wallet, token, chain) => {
                                    trackPayEvent({
                                        chainId: chain.id,
                                        client: props.client,
                                        event: "choose_payment_method_token",
                                        fromToken: isNativeToken(token) ? undefined : token.address,
                                        toChainId: props.toChain.id,
                                        toToken: isNativeToken(props.toToken)
                                            ? undefined
                                            : props.toToken.address,
                                        walletAddress: activeAccount?.address,
                                        walletType: activeWallet?.id,
                                    });
                                    props.onSelectToken(wallet, token, chain);
                                }, wallet: wallet }, w.id));
                        }), filteredWallets.length > 0 && _jsx(TextDivider, { text: "OR" }), _jsx(Button, { bg: "tertiaryBg", fullWidth: true, onClick: () => {
                                trackPayEvent({
                                    client: props.client,
                                    event: "choose_payment_method_another_wallet",
                                    toChainId: props.toChain.id,
                                    toToken: isNativeToken(props.toToken)
                                        ? undefined
                                        : props.toToken.address,
                                    walletAddress: activeAccount?.address,
                                    walletType: activeWallet?.id,
                                });
                                props.onConnect();
                            }, style: {
                                border: `1px solid ${theme.colors.borderColor}`,
                                padding: spacing.sm,
                            }, variant: "secondary", children: _jsxs(Container, { center: "y", color: "secondaryIconColor", expand: true, flex: "row", gap: "sm", children: [_jsx(OutlineWalletIcon, { size: iconSize.md }), _jsx(Text, { color: "primaryText", size: "sm", children: "Pay with another wallet" })] }) }), props.fiatSupported && (_jsx(Button, { bg: "tertiaryBg", fullWidth: true, onClick: () => {
                                trackPayEvent({
                                    client: props.client,
                                    event: "choose_payment_method_with_card",
                                    toChainId: props.toChain.id,
                                    toToken: isNativeToken(props.toToken)
                                        ? undefined
                                        : props.toToken.address,
                                    walletAddress: activeAccount?.address,
                                    walletType: activeWallet?.id,
                                });
                                props.onPayWithFiat();
                            }, style: {
                                border: `1px solid ${theme.colors.borderColor}`,
                                padding: spacing.sm,
                            }, variant: "secondary", children: _jsxs(Container, { center: "y", color: "secondaryIconColor", expand: true, flex: "row", gap: "sm", children: [_jsx(CardStackIcon, { height: iconSize.md, width: iconSize.md }), _jsx(Text, { color: "primaryText", size: "sm", children: "Pay with card" })] }) }))] }) })] }));
}
function WalletRowWithBalances(props) {
    const theme = useCustomTheme();
    const displayedBalances = props.balances;
    const activeAccount = useActiveAccount();
    const { disconnect } = useDisconnect();
    const isActiveAccount = activeAccount?.address === props.address;
    return (_jsxs(Container, { flex: "column", style: {
            border: `1px solid ${theme.colors.borderColor}`,
            borderRadius: radius.lg,
        }, children: [_jsxs(Container, { bg: "tertiaryBg", flex: "row", gap: "sm", style: {
                    borderBottom: `1px solid ${theme.colors.borderColor}`,
                    borderTopLeftRadius: radius.lg,
                    borderTopRightRadius: radius.lg,
                    justifyContent: "space-between",
                    padding: spacing.sm,
                    paddingRight: spacing.xs,
                }, children: [_jsx(WalletRow, { ...props }), !isActiveAccount && (_jsx(Button, { onClick: () => disconnect(props.wallet), style: {
                            color: theme.colors.secondaryText,
                            padding: spacing.xxs,
                        }, variant: "ghost", children: _jsx(Cross2Icon, { height: iconSize.sm, width: iconSize.sm }) }))] }), _jsx(Container, { flex: "column", children: props.balances.length > 0 ? (displayedBalances.map((b, idx) => (_jsx(TokenBalanceRow, { client: props.client, onClick: () => props.onClick(props.wallet, b.token, b.chain), style: {
                        borderBottom: idx === displayedBalances.length - 1
                            ? "none"
                            : `1px solid ${theme.colors.borderColor}`,
                        borderBottomLeftRadius: idx === displayedBalances.length - 1 ? radius.lg : 0,
                        borderBottomRightRadius: idx === displayedBalances.length - 1 ? radius.lg : 0,
                        borderTopLeftRadius: 0,
                        borderTopRightRadius: 0,
                    }, tokenBalance: b, wallet: props.wallet }, `${b.token.address}-${b.chain.id}`)))) : (_jsx(Container, { style: { padding: spacing.sm }, children: _jsx(Text, { color: "secondaryText", size: "sm", children: "Insufficient Funds" }) })) })] }));
}
function TokenBalanceRow(props) {
    const { tokenBalance, wallet, onClick, client, style } = props;
    const chainInfo = useChainName(tokenBalance.chain);
    return (_jsxs(StyledButton, { onClick: () => onClick(tokenBalance.token, wallet), style: {
            ...style,
            display: "flex",
            justifyContent: "space-between",
            minWidth: 0, // Needed for text truncation to work
        }, variant: "secondary", children: [_jsxs(Container, { center: "y", flex: "row", gap: "sm", style: {
                    flex: "1 1 50%",
                    flexWrap: "nowrap",
                    maxWidth: "50%",
                    minWidth: 0,
                    overflow: "hidden",
                }, children: [_jsx(TokenIcon, { chain: tokenBalance.chain, client: client, size: "md", token: tokenBalance.token }), _jsxs(Container, { flex: "column", gap: "4xs", style: { minWidth: 0 }, children: [_jsx(Text, { color: "primaryText", size: "xs", style: {
                                    overflow: "hidden",
                                    textOverflow: "ellipsis",
                                    whiteSpace: "nowrap",
                                }, children: tokenBalance.token.symbol }), chainInfo && (_jsx(Text, { size: "xs", style: {
                                    overflow: "hidden",
                                    textOverflow: "ellipsis",
                                    whiteSpace: "nowrap",
                                }, children: chainInfo.name }))] })] }), _jsxs(Container, { center: "y", color: "secondaryText", flex: "row", gap: "4xs", style: {
                    flex: "1 1 50%",
                    flexWrap: "nowrap",
                    justifyContent: "flex-end",
                    maxWidth: "50%",
                    minWidth: 0,
                }, children: [_jsxs(Container, { color: "secondaryText", flex: "column", gap: "4xs", style: {
                            alignItems: "flex-end",
                            minWidth: 0,
                            overflow: "hidden",
                        }, children: [_jsx(Text, { color: "primaryText", size: "xs", style: {
                                    overflow: "hidden",
                                    textOverflow: "ellipsis",
                                    whiteSpace: "nowrap",
                                }, children: formatTokenBalance(tokenBalance.balance, true, 2) }), _jsx(FiatValue, { chain: tokenBalance.chain, client: client, size: "xs", token: tokenBalance.token, tokenAmount: tokenBalance.balance.displayValue })] }), _jsx(ChevronRightIcon, { height: iconSize.md, style: { flexShrink: 0 }, width: iconSize.md })] })] }));
}
const StyledButton = /* @__PURE__ */ styled(Button)((props) => {
    const theme = useCustomTheme();
    return {
        "&:hover": {
            background: theme.colors.secondaryButtonBg,
            transform: "scale(1.01)",
        },
        background: "transparent",
        flexDirection: "row",
        flexWrap: "nowrap",
        gap: spacing.sm,
        justifyContent: "space-between",
        padding: spacing.sm,
        paddingRight: spacing.xs,
        transition: "background 200ms ease, transform 150ms ease",
        ...props.style,
    };
});
