// thirdweb/dist/cjs/wallets/__generated__/wallet/io.getclave/index.js
"use strict";
// This file is auto-generated by the `scripts/wallets/generate.ts` script.
// Do not modify this file manually.
Object.defineProperty(exports, "__esModule", { value: true });
exports.wallet = void 0;
exports.wallet = {
    app: {
        android: "https://play.google.com/store/apps/details?id=com.clave.mobile",
        browser: null,
        chrome: null,
        edge: null,
        firefox: null,
        ios: "https://apps.apple.com/gr/app/clave-smart-wallet/id6449253761",
        linux: null,
        mac: "https://apps.apple.com/gr/app/clave-smart-wallet/id6449253761",
        opera: null,
        safari: null,
        windows: null,
    },
    desktop: {
        native: "clave://link/wc/",
        universal: "https://getclave.io/link",
    },
    homepage: "https://getclave.io",
    id: "io.getclave",
    image_id: "f1c538df-15d9-4448-542f-b7b358e95d00",
    mobile: {
        native: "clave://link/wc/",
        universal: "https://getclave.io/link/wc",
    },
    name: "Clave",
    rdns: null,
};
