// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/swap/SwapScreenContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwapScreenContent = SwapScreenContent;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_query_1 = require("@tanstack/react-query");
const react_1 = require("react");
const pay_js_1 = require("../../../../../../../analytics/track/pay.js");
const utils_js_1 = require("../../../../../../../chains/utils.js");
const addresses_js_1 = require("../../../../../../../constants/addresses.js");
const contract_js_1 = require("../../../../../../../contract/contract.js");
const allowance_js_1 = require("../../../../../../../extensions/erc20/__generated__/IERC20/read/allowance.js");
const useWalletBalance_js_1 = require("../../../../../../core/hooks/others/useWalletBalance.js");
const useBuyWithCryptoQuote_js_1 = require("../../../../../../core/hooks/pay/useBuyWithCryptoQuote.js");
const errors_js_1 = require("../../../../../utils/errors.js");
const basic_js_1 = require("../../../../components/basic.js");
const buttons_js_1 = require("../../../../components/buttons.js");
const Drawer_js_1 = require("../../../../components/Drawer.js");
const Spacer_js_1 = require("../../../../components/Spacer.js");
const Spinner_js_1 = require("../../../../components/Spinner.js");
const text_js_1 = require("../../../../components/text.js");
const TokenSymbol_js_1 = require("../../../../components/token/TokenSymbol.js");
const nativeToken_js_1 = require("../../nativeToken.js");
const EstimatedTimeAndFees_js_1 = require("../EstimatedTimeAndFees.js");
const Fees_js_1 = require("./Fees.js");
const PayWithCrypto_js_1 = require("./PayWithCrypto.js");
function SwapScreenContent(props) {
    const { setScreen, payer, client, toChain, tokenAmount, toToken, fromChain, fromToken, payOptions, disableTokenSelection, } = props;
    const defaultRecipientAddress = props.payOptions?.paymentInfo?.sellerAddress;
    const receiverAddress = defaultRecipientAddress || props.activeAccount.address;
    const { drawerRef, drawerOverlayRef, isOpen, setIsOpen } = (0, Drawer_js_1.useDrawer)();
    const [drawerScreen, setDrawerScreen] = (0, react_1.useState)("fees");
    const fromTokenBalanceQuery = (0, useWalletBalance_js_1.useWalletBalance)({
        address: payer.account.address,
        chain: fromChain,
        client,
        tokenAddress: (0, nativeToken_js_1.isNativeToken)(fromToken) ? undefined : fromToken?.address,
    }, {
        enabled: !!fromChain && !!fromToken,
    });
    const fromTokenId = (0, nativeToken_js_1.isNativeToken)(fromToken)
        ? addresses_js_1.NATIVE_TOKEN_ADDRESS
        : fromToken?.address?.toLowerCase();
    const toTokenId = (0, nativeToken_js_1.isNativeToken)(toToken)
        ? addresses_js_1.NATIVE_TOKEN_ADDRESS
        : toToken.address.toLowerCase();
    const swapRequired = !!tokenAmount &&
        !!fromChain &&
        !!fromTokenId &&
        !(fromChain?.id === toChain.id && fromTokenId === toTokenId);
    const quoteParams = fromChain && fromToken && swapRequired
        ? {
            client,
            // wallets
            fromAddress: payer.account.address,
            // from
            fromChainId: fromChain.id,
            fromTokenAddress: (0, nativeToken_js_1.isNativeToken)(fromToken)
                ? addresses_js_1.NATIVE_TOKEN_ADDRESS
                : fromToken.address,
            paymentLinkId: props.paymentLinkId,
            purchaseData: payOptions.purchaseData,
            toAddress: receiverAddress,
            toAmount: tokenAmount,
            // to
            toChainId: toChain.id,
            toTokenAddress: (0, nativeToken_js_1.isNativeToken)(toToken)
                ? addresses_js_1.NATIVE_TOKEN_ADDRESS
                : toToken.address,
        }
        : undefined;
    const quoteQuery = (0, useBuyWithCryptoQuote_js_1.useBuyWithCryptoQuote)(quoteParams, {
        gcTime: 30 * 1000,
        refetchInterval: 30 * 1000,
        // refetch every 30 seconds
        staleTime: 30 * 1000,
    });
    const allowanceQuery = (0, react_query_1.useQuery)({
        enabled: !!quoteQuery.data?.approvalData,
        queryFn: () => {
            if (!quoteQuery.data?.approvalData) {
                return null;
            }
            return (0, allowance_js_1.allowance)({
                contract: (0, contract_js_1.getContract)({
                    address: quoteQuery.data.swapDetails.fromToken.tokenAddress,
                    chain: (0, utils_js_1.getCachedChain)(quoteQuery.data.swapDetails.fromToken.chainId),
                    client: props.client,
                }),
                owner: props.payer.account.address,
                spender: quoteQuery.data.approvalData.spenderAddress,
            });
        },
        queryKey: [
            "allowance",
            payer.account.address,
            quoteQuery.data?.approvalData,
        ],
        refetchOnMount: true,
    });
    const sourceTokenAmount = swapRequired
        ? quoteQuery.data?.swapDetails.fromAmount
        : tokenAmount;
    const isNotEnoughBalance = !!sourceTokenAmount &&
        !!fromTokenBalanceQuery.data &&
        Number(fromTokenBalanceQuery.data.displayValue) < Number(sourceTokenAmount);
    const disableContinue = !fromChain ||
        !fromToken ||
        (swapRequired && !quoteQuery.data) ||
        isNotEnoughBalance ||
        allowanceQuery.isLoading;
    const errorMsg = !quoteQuery.isLoading && quoteQuery.error
        ? (0, errors_js_1.getErrorMessage)(quoteQuery.error)
        : undefined;
    function showSwapFlow() {
        if ((props.payOptions.mode === "direct_payment" ||
            props.payOptions.mode === "fund_wallet") &&
            !isNotEnoughBalance &&
            !swapRequired) {
            // same currency, just direct transfer
            setScreen({
                id: "transfer-flow",
            });
        }
        else if (props.payOptions.mode === "transaction" &&
            !isNotEnoughBalance &&
            !swapRequired) {
            if (payer.account.address !== receiverAddress) {
                // needs transfer from another wallet before executing the transaction
                setScreen({
                    id: "transfer-flow",
                });
            }
            else {
                // has enough balance to just do the transaction directly
                props.onDone();
            }
            return;
        }
        if (!quoteQuery.data) {
            return;
        }
        setScreen({
            approvalAmount: allowanceQuery.data ?? undefined,
            id: "swap-flow",
            quote: quoteQuery.data,
        });
    }
    function showFees() {
        if (!quoteQuery.data) {
            return;
        }
        setIsOpen(true);
        setDrawerScreen("fees");
    }
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { animate: "fadein", flex: "column", gap: "lg", children: [isOpen && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Drawer_js_1.DrawerOverlay, { ref: drawerOverlayRef }), (0, jsx_runtime_1.jsx)(Drawer_js_1.Drawer, { close: () => setIsOpen(false), ref: drawerRef, children: drawerScreen === "fees" && quoteQuery.data && ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "lg", children: "Fees" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(Fees_js_1.SwapFees, { quote: quoteQuery.data })] })) })] })), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "sm", children: [(0, jsx_runtime_1.jsxs)(basic_js_1.Container, { center: "y", flex: "row", gap: "xxs", children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "sm", children: "Pay with" }), fromToken && fromChain ? ((0, jsx_runtime_1.jsx)(TokenSymbol_js_1.TokenSymbol, { chain: fromChain, color: "secondaryText", size: "sm", token: fromToken })) : ("crypto")] }), (0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(PayWithCrypto_js_1.PayWithCryptoQuoteInfo, { chain: fromChain, client: client, freezeChainAndTokenSelection: disableTokenSelection, isLoading: quoteQuery.isLoading && !sourceTokenAmount, onSelectToken: props.showFromTokenSelector, payerAccount: props.payer.account, swapRequired: swapRequired, token: fromToken, value: sourceTokenAmount || "" }), swapRequired && fromChain && fromToken && ((0, jsx_runtime_1.jsx)(EstimatedTimeAndFees_js_1.EstimatedTimeAndFees, { estimatedSeconds: quoteQuery.data?.swapDetails.estimated.durationSeconds, onViewFees: showFees, quoteIsLoading: quoteQuery.isLoading }))] }), errorMsg && ((0, jsx_runtime_1.jsx)("div", { children: (0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, color: "danger", multiline: true, size: "xs", children: errorMsg.title }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, multiline: true, size: "xs", children: errorMsg.message })] }) })), !errorMsg && isNotEnoughBalance && ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, color: "danger", multiline: true, size: "xs", children: "Insufficient Funds" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, multiline: true, size: "xs", children: "Select another token or pay with card." })] }))] }), isNotEnoughBalance || errorMsg ? ((0, jsx_runtime_1.jsx)(buttons_js_1.Button, { fullWidth: true, onClick: () => props.showFromTokenSelector(), variant: "accent", children: "Pay with another token" })) : ((0, jsx_runtime_1.jsx)(buttons_js_1.Button, { "data-disabled": disableContinue, disabled: disableContinue, fullWidth: true, gap: "xs", onClick: async () => {
                    if (!disableContinue) {
                        if (props.payer.wallet.getChain()?.id !== fromChain?.id) {
                            await props.payer.wallet.switchChain(fromChain);
                        }
                        showSwapFlow();
                        (0, pay_js_1.trackPayEvent)({
                            chainId: fromChain.id,
                            client: client,
                            event: "confirm_swap_quote",
                            fromToken: (0, nativeToken_js_1.isNativeToken)(fromToken)
                                ? undefined
                                : fromToken.address,
                            toChainId: toChain.id,
                            toToken: (0, nativeToken_js_1.isNativeToken)(toToken) ? undefined : toToken.address,
                            walletAddress: payer.account.address,
                            walletType: payer.wallet.id,
                        });
                    }
                }, variant: disableContinue ? "outline" : "accent", children: quoteQuery.isLoading ? ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: ["Getting price quote", (0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { size: "sm" })] })) : ("Continue") }))] }));
}
