// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/PayWIthCreditCard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayWithCreditCard = PayWithCreditCard;
const jsx_runtime_1 = require("react/jsx-runtime");
const styled_1 = require("@emotion/styled");
const react_icons_1 = require("@radix-ui/react-icons");
const formatNumber_js_1 = require("../../../../../../utils/formatNumber.js");
const index_js_1 = require("../../../../../core/design-system/index.js");
const basic_js_1 = require("../../../components/basic.js");
const buttons_js_1 = require("../../../components/buttons.js");
const Skeleton_js_1 = require("../../../components/Skeleton.js");
const text_js_1 = require("../../../components/text.js");
const currencies_js_1 = require("./fiat/currencies.js");
/**
 * Shows an amount "value" and renders the selected token and chain
 * It also renders the buttons to select the token and chain
 * It also renders the balance of active wallet for the selected token in selected chain
 * @internal
 */
function PayWithCreditCard(props) {
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { bg: "tertiaryBg", borderColor: "borderColor", flex: "row", style: {
            alignItems: "center",
            borderBottom: "none",
            borderBottomLeftRadius: 0,
            borderBottomRightRadius: 0,
            borderRadius: index_js_1.radius.md,
            borderStyle: "solid",
            borderWidth: "1px",
            flexWrap: "nowrap",
            justifyContent: "space-between",
        }, children: [(0, jsx_runtime_1.jsxs)(CurrencyButton, { gap: "sm", onClick: props.onSelectCurrency, style: {
                    justifyContent: "flex-start",
                    minHeight: "64px",
                    minWidth: "50%",
                }, variant: "ghost", children: [(0, currencies_js_1.getFiatIcon)(props.currency, "md"), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { center: "y", color: "secondaryText", flex: "row", gap: "xxs", children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", children: props.currency.shorthand }), (0, jsx_runtime_1.jsx)(react_icons_1.ChevronDownIcon, { height: index_js_1.iconSize.sm, width: index_js_1.iconSize.sm })] })] }), (0, jsx_runtime_1.jsx)("div", { style: {
                    alignItems: "flex-end",
                    display: "flex",
                    flexDirection: "column",
                    flexGrow: 1,
                    flexShrink: 1,
                    gap: index_js_1.spacing.xxs,
                    justifyContent: "center",
                    overflow: "hidden",
                    paddingRight: index_js_1.spacing.sm,
                    textOverflow: "ellipsis",
                    whiteSpace: "nowrap",
                }, children: props.isLoading ? ((0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { height: index_js_1.fontSize.lg, width: "100px" })) : ((0, jsx_runtime_1.jsx)(text_js_1.Text, { color: props.value ? "primaryText" : "secondaryText", size: "lg", children: props.value
                        ? `${props.currency.symbol}${(0, formatNumber_js_1.formatNumber)(Number(props.value), 2)}`
                        : "--" })) })] }));
}
const CurrencyButton = /* @__PURE__ */ (0, styled_1.default)(buttons_js_1.Button)(() => {
    return {
        "&[disabled]:hover": {
            borderColor: "transparent",
        },
    };
});
