// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/fiat/CurrencySelection.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import styled from "@emotion/styled";
import { useCustomTheme } from "../../../../../../core/design-system/CustomThemeProvider.js";
import { spacing } from "../../../../../../core/design-system/index.js";
import { Container, Line, ModalHeader } from "../../../../components/basic.js";
import { Button } from "../../../../components/buttons.js";
import { Spacer } from "../../../../components/Spacer.js";
import { Text } from "../../../../components/text.js";
import { currencies, getFiatIcon } from "./currencies.js";
export function CurrencySelection(props) {
    return (_jsxs(Container, { children: [_jsx(Container, { p: "lg", children: _jsx(ModalHeader, { onBack: props.onBack, title: "Pay with" }) }), _jsx(Line, {}), _jsx(Spacer, { y: "lg" }), _jsx(Container, { flex: "column", gap: "xs", px: "lg", children: currencies.map((c) => {
                    return (_jsxs(SelectCurrencyButton, { fullWidth: true, gap: "sm", onClick: () => props.onSelect(c), variant: "secondary", children: [getFiatIcon(c, "lg"), _jsxs(Container, { flex: "column", gap: "xxs", children: [_jsx(Text, { color: "primaryText", children: c.shorthand }), _jsx(Text, { size: "sm", children: c.name })] })] }, c.shorthand));
                }) }), _jsx(Spacer, { y: "lg" })] }));
}
const SelectCurrencyButton = /* @__PURE__ */ styled(Button)(() => {
    const theme = useCustomTheme();
    return {
        "&:hover": {
            background: theme.colors.secondaryButtonBg,
            transform: "scale(1.01)",
        },
        background: theme.colors.tertiaryBg,
        gap: spacing.sm,
        justifyContent: "flex-start",
        padding: spacing.sm,
        transition: "background 200ms ease, transform 150ms ease",
    };
});
