// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/fiat/FiatSteps.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Cross1Icon } from "@radix-ui/react-icons";
import { iconSize, radius, spacing, } from "../../../../../../core/design-system/index.js";
import { Container } from "../../../../components/basic.js";
import { Spinner } from "../../../../components/Spinner.js";
import { Text } from "../../../../components/text.js";
import { StepIcon } from "../Stepper.js";
export function StepContainer(props) {
    let color = "borderColor";
    let text;
    if (props.state === "pending") {
        text = "Pending";
        color = "accentText";
    }
    else if (props.state === "actionRequired") {
        color = "accentText";
    }
    else if (props.state === "completed") {
        text = "Completed";
        color = "success";
    }
    else if (props.state === "failed") {
        color = "danger";
        text = "Failed";
    }
    return (_jsxs(Container, { bg: "tertiaryBg", borderColor: color === "success" ? "borderColor" : color, px: "md", py: "sm", style: {
            alignItems: "flex-start",
            borderRadius: radius.lg,
            borderStyle: "solid",
            borderWidth: "1px",
            position: "relative",
            ...props.style,
        }, children: [props.children, _jsxs("div", { style: {
                    alignItems: "center",
                    display: "flex",
                    gap: spacing.xs,
                    position: "absolute",
                    right: spacing.xs,
                    top: spacing.xs,
                }, children: [props.state && text && (_jsx(Text, { color: color, size: "xs", children: text })), (props.state === "actionRequired" || props.state === "completed") && (_jsx(StepIcon, { isActive: props.state === "actionRequired", isDone: props.state === "completed" })), props.state === "pending" && _jsx(Spinner, { color: "accentText", size: "sm" }), props.state === "failed" && (_jsx(Container, { center: "both", color: "danger", flex: "row", children: _jsx(Cross1Icon, { height: iconSize.sm, width: iconSize.sm }) }))] })] }));
}
