// thirdweb/dist/esm/react/web/ui/components/token/TokenRow.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import styled from "@emotion/styled";
import { ChevronDownIcon } from "@radix-ui/react-icons";
import { formatNumber } from "../../../../../utils/formatNumber.js";
import { useCustomTheme } from "../../../../core/design-system/CustomThemeProvider.js";
import { fontSize, iconSize, spacing, } from "../../../../core/design-system/index.js";
import { useChainName } from "../../../../core/hooks/others/useChainQuery.js";
import { PayTokenIcon } from "../../ConnectWallet/screens/Buy/PayTokenIcon.js";
import { Container } from "../basic.js";
import { Button } from "../buttons.js";
import { Skeleton } from "../Skeleton.js";
import { Text } from "../text.js";
import { TokenSymbol } from "./TokenSymbol.js";
export function TokenRow(props) {
    const { name } = useChainName(props.chain);
    if (!props.token || !props.chain) {
        return (_jsxs(Button, { fullWidth: true, gap: "xxs", onClick: props.onSelectToken, style: {
                fontSize: fontSize.sm,
                justifyContent: "space-between",
                paddingBottom: spacing.md,
                paddingTop: spacing.md,
                ...props.style,
            }, variant: "secondary", children: [_jsx(Text, { color: "primaryText", size: "sm", children: "Select payment token" }), _jsx(Container, { color: "primaryText", children: _jsx(ChevronDownIcon, { height: iconSize.sm, width: iconSize.sm }) })] }));
    }
    return (_jsxs(TokenButton, { disabled: props.freezeChainAndToken, fullWidth: true, gap: "xxs", onClick: props.onSelectToken, style: {
            fontSize: fontSize.sm,
            ...props.style,
        }, variant: "secondary", children: [_jsxs(Container, { center: "y", flex: "row", gap: "sm", children: [_jsx(PayTokenIcon, { chain: props.chain, client: props.client, size: "md", token: props.token }), _jsxs(Container, { flex: "column", gap: "4xs", children: [_jsx(Container, { flex: "column", gap: "4xs", children: props.isLoading ? (_jsx(Skeleton, { color: "borderColor", height: fontSize.md, width: "120px" })) : props.value ? (_jsxs(Container, { center: "y", color: "primaryText", flex: "row", gap: "xxs", children: [_jsx(Text, { color: props.value ? "primaryText" : "secondaryText", size: "md", children: formatNumber(Number(props.value), 6) || "" }), _jsx(TokenSymbol, { chain: props.chain, color: "secondaryText", size: "sm", token: props.token })] })) : (_jsx(TokenSymbol, { chain: props.chain, size: "sm", token: props.token })) }), name ? (_jsx(Text, { color: "secondaryText", size: "xs", children: name })) : (_jsx(Skeleton, { height: fontSize.xs, width: "90px" }))] })] }), !props.freezeChainAndToken && (_jsx(Container, { color: "primaryText", children: _jsx(ChevronDownIcon, { height: iconSize.sm, width: iconSize.sm }) }))] }));
}
const TokenButton = /* @__PURE__ */ styled(Button)(() => {
    const theme = useCustomTheme();
    return {
        background: theme.colors.tertiaryBg,
        border: `1px solid ${theme.colors.borderColor}`,
        justifyContent: "space-between",
        padding: spacing.sm,
        transition: "background 0.3s",
    };
});
