// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/fiat/FiatScreenContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FiatScreenContent = FiatScreenContent;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const pay_js_1 = require("../../../../../../../analytics/track/pay.js");
const addresses_js_1 = require("../../../../../../../constants/addresses.js");
const commonTypes_js_1 = require("../../../../../../../pay/utils/commonTypes.js");
const useBuyWithFiatQuote_js_1 = require("../../../../../../core/hooks/pay/useBuyWithFiatQuote.js");
const storage_js_1 = require("../../../../../../core/utils/storage.js");
const errors_js_1 = require("../../../../../utils/errors.js");
const basic_js_1 = require("../../../../components/basic.js");
const buttons_js_1 = require("../../../../components/buttons.js");
const Drawer_js_1 = require("../../../../components/Drawer.js");
const Spacer_js_1 = require("../../../../components/Spacer.js");
const Spinner_js_1 = require("../../../../components/Spinner.js");
const text_js_1 = require("../../../../components/text.js");
const nativeToken_js_1 = require("../../nativeToken.js");
const EstimatedTimeAndFees_js_1 = require("../EstimatedTimeAndFees.js");
const PayProviderSelection_js_1 = require("../PayProviderSelection.js");
const PayWIthCreditCard_js_1 = require("../PayWIthCreditCard.js");
const Fees_js_1 = require("../swap/Fees.js");
const Providers_js_1 = require("./Providers.js");
function FiatScreenContent(props) {
    const { toToken, tokenAmount, payer, client, setScreen, toChain, showCurrencySelector, selectedCurrency, paymentLinkId, } = props;
    const defaultRecipientAddress = props.payOptions?.paymentInfo?.sellerAddress;
    const receiverAddress = defaultRecipientAddress || props.payer.account.address;
    const { drawerRef, drawerOverlayRef, isOpen, setIsOpen } = (0, Drawer_js_1.useDrawer)();
    const [drawerScreen, setDrawerScreen] = (0, react_1.useState)("fees");
    const buyWithFiatOptions = props.payOptions.buyWithFiat;
    const [preferredProvider, setPreferredProvider] = (0, react_1.useState)(buyWithFiatOptions !== false
        ? buyWithFiatOptions?.preferredProvider ||
            (localStorage.getItem(storage_js_1.PREFERRED_FIAT_PROVIDER_STORAGE_KEY) ??
                undefined)
        : undefined);
    const supportedProviders = (() => {
        if (!buyWithFiatOptions)
            return [...commonTypes_js_1.FiatProviders];
        const options = buyWithFiatOptions?.supportedProviders ?? [];
        const optionsWithPreferred = options.length > 0
            ? new Set([
                ...(preferredProvider ? [preferredProvider] : []),
                ...options,
            ])
            : commonTypes_js_1.FiatProviders;
        return Array.from(optionsWithPreferred);
    })();
    const fiatQuoteQuery = (0, useBuyWithFiatQuote_js_1.useBuyWithFiatQuote)(buyWithFiatOptions !== false && tokenAmount
        ? {
            client,
            fromAddress: payer.account.address,
            fromCurrencySymbol: selectedCurrency.shorthand,
            isTestMode: buyWithFiatOptions?.testMode,
            onrampChainId: buyWithFiatOptions?.onrampChainId,
            onrampTokenAddress: buyWithFiatOptions?.onrampTokenAddress,
            paymentLinkId: paymentLinkId,
            preferredProvider: preferredProvider ?? supportedProviders[0],
            purchaseData: props.payOptions.purchaseData,
            toAddress: receiverAddress,
            toAmount: tokenAmount,
            toChainId: toChain.id,
            toTokenAddress: (0, nativeToken_js_1.isNativeToken)(toToken)
                ? addresses_js_1.NATIVE_TOKEN_ADDRESS
                : toToken.address,
        }
        : undefined);
    function handleSubmit() {
        if (!fiatQuoteQuery.data) {
            return;
        }
        setScreen({
            id: "fiat-flow",
            quote: fiatQuoteQuery.data,
        });
    }
    function showFees() {
        if (!fiatQuoteQuery.data) {
            return;
        }
        setDrawerScreen("fees");
        setIsOpen(true);
    }
    function showProviders() {
        setDrawerScreen("providers");
        setIsOpen(true);
    }
    const disableSubmit = !fiatQuoteQuery.data;
    const errorMsg = !fiatQuoteQuery.isLoading && fiatQuoteQuery.error
        ? (0, errors_js_1.getErrorMessage)(fiatQuoteQuery.error)
        : undefined;
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { animate: "fadein", flex: "column", gap: "lg", children: [isOpen && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Drawer_js_1.DrawerOverlay, { ref: drawerOverlayRef }), (0, jsx_runtime_1.jsxs)(Drawer_js_1.Drawer, { close: () => setIsOpen(false), ref: drawerRef, children: [drawerScreen === "fees" && fiatQuoteQuery.data && ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "lg", children: "Fees" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(Fees_js_1.FiatFees, { quote: fiatQuoteQuery.data })] })), drawerScreen === "providers" && ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "lg", children: "Providers" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(Providers_js_1.Providers, { onSelect: (provider) => {
                                            setPreferredProvider(provider);
                                            // save the pref in local storage
                                            localStorage.setItem(storage_js_1.PREFERRED_FIAT_PROVIDER_STORAGE_KEY, provider);
                                            setIsOpen(false);
                                        }, preferredProvider: preferredProvider || fiatQuoteQuery.data?.provider, supportedProviders: supportedProviders })] }))] })] })), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "sm", children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "sm", children: "Pay with card" }), (0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(PayWIthCreditCard_js_1.PayWithCreditCard, { client: client, currency: selectedCurrency, isLoading: fiatQuoteQuery.isLoading, onSelectCurrency: showCurrencySelector, value: fiatQuoteQuery.data?.fromCurrencyWithFees.amount }), (0, jsx_runtime_1.jsx)(PayProviderSelection_js_1.PayProviderSelection, { onShowProviders: showProviders, preferredProvider: preferredProvider, quotedProvider: fiatQuoteQuery.data?.provider, supportedProviders: supportedProviders }), (0, jsx_runtime_1.jsx)(EstimatedTimeAndFees_js_1.EstimatedTimeAndFees, { estimatedSeconds: fiatQuoteQuery.data?.estimatedDurationSeconds, onViewFees: showFees, quoteIsLoading: fiatQuoteQuery.isLoading })] }), errorMsg && ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, color: "danger", multiline: true, size: "xs", children: errorMsg.title }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, multiline: true, size: "xs", children: errorMsg.message })] }))] }), (0, jsx_runtime_1.jsx)(buttons_js_1.Button, { "data-disabled": disableSubmit, disabled: disableSubmit, fullWidth: true, gap: "xs", onClick: () => {
                    (0, pay_js_1.trackPayEvent)({
                        client: client,
                        event: "confirm_onramp_quote",
                        toChainId: toChain.id,
                        toToken: (0, nativeToken_js_1.isNativeToken)(toToken) ? undefined : toToken.address,
                        walletAddress: payer.account.address,
                        walletType: payer.wallet.id,
                    });
                    handleSubmit();
                }, variant: disableSubmit ? "outline" : "accent", children: fiatQuoteQuery.isLoading ? ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: ["Getting price quote", (0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { size: "sm" })] })) : ("Continue") })] }));
}
