// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/fiat/FiatSteps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepContainer = StepContainer;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_icons_1 = require("@radix-ui/react-icons");
const index_js_1 = require("../../../../../../core/design-system/index.js");
const basic_js_1 = require("../../../../components/basic.js");
const Spinner_js_1 = require("../../../../components/Spinner.js");
const text_js_1 = require("../../../../components/text.js");
const Stepper_js_1 = require("../Stepper.js");
function StepContainer(props) {
    let color = "borderColor";
    let text;
    if (props.state === "pending") {
        text = "Pending";
        color = "accentText";
    }
    else if (props.state === "actionRequired") {
        color = "accentText";
    }
    else if (props.state === "completed") {
        text = "Completed";
        color = "success";
    }
    else if (props.state === "failed") {
        color = "danger";
        text = "Failed";
    }
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { bg: "tertiaryBg", borderColor: color === "success" ? "borderColor" : color, px: "md", py: "sm", style: {
            alignItems: "flex-start",
            borderRadius: index_js_1.radius.lg,
            borderStyle: "solid",
            borderWidth: "1px",
            position: "relative",
            ...props.style,
        }, children: [props.children, (0, jsx_runtime_1.jsxs)("div", { style: {
                    alignItems: "center",
                    display: "flex",
                    gap: index_js_1.spacing.xs,
                    position: "absolute",
                    right: index_js_1.spacing.xs,
                    top: index_js_1.spacing.xs,
                }, children: [props.state && text && ((0, jsx_runtime_1.jsx)(text_js_1.Text, { color: color, size: "xs", children: text })), (props.state === "actionRequired" || props.state === "completed") && ((0, jsx_runtime_1.jsx)(Stepper_js_1.StepIcon, { isActive: props.state === "actionRequired", isDone: props.state === "completed" })), props.state === "pending" && (0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { color: "accentText", size: "sm" }), props.state === "failed" && ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { center: "both", color: "danger", flex: "row", children: (0, jsx_runtime_1.jsx)(react_icons_1.Cross1Icon, { height: index_js_1.iconSize.sm, width: index_js_1.iconSize.sm }) }))] })] }));
}
